lib/Conversion/StandardToHandshake/StandardToHandshake.cpp

```diff
@@ -1660,6 +1660,78 @@ struct HandshakeAnalysisPass
 };
 } // namespace
 
+// This pass iteratively checks the use count of each value, and reformats the
+// MLIR such that every value has exact one use. If a value has multiple uses, a
+// fork is inserted. If a value has no use, a sink is inserted.
+void insertOpForExactOneUse(Value value, handshake::FuncOp funcOp,
+                            OpBuilder &rewriter) {
+
+  if (value.use_empty())
+    rewriter.create<SinkOp>(funcOp.getLoc(), value);
+  else {
+    auto uses = value.getUses();
+    auto outs = std::distance(cbegin(uses), cend(uses));
+    std::vector<Operation *> ops;
+    std::vector<unsigned> idx;
+    for (auto &use : uses) {
+      ops.push_back(use.getOwner());
+      idx.push_back(use.getOperandNumber());
+    }
+
+    // There is an unkown issue when forking the function arguments directly
+    // The temporary solution to to create merge followed by the fork, and
+    // remove the merge after creating fork
+    rewriter.setInsertionPoint(
+        &(funcOp.getOperation()->getRegions().begin()->begin()->back()));
+    auto tempMerge = rewriter.create<circt::handshake::MergeOp>(
+        funcOp.getLoc(), value.getUses().begin()->get(), 1);
+    auto newOp = rewriter.create<circt::handshake::ForkOp>(
+        funcOp.getLoc(), tempMerge.getResult(), outs);
+
+    for (auto i = 0; i < outs; i++)
```

Comment:
I'd recommend ditching the temporary vectors above and just iterating through uses here and setting it directly. If necessary, make the merge a case to avoid. I don't know if it's safe to modify uses while iterating through them, so you may have to make a copy and iterate over that.